sdk/typescript/src/experimental/types.ts

```diff
@@ -0,0 +1,253 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/**
+ * Types and Interface definitions for the SDK.
+ *
+ * They provide the foundation for the TypeScript system which mimics
+ * the underlying BCS structs.
+ *
+ * @module sui-types
+ */
+
+import { BCS, decodeStr, encodeStr } from "./bcs";
+
+// This buddy collects definitions for BCS.
+const initializers: Function[] = [];
+
+initializers.push((b: typeof BCS) =>
+  b
+    .registerVectorType("vector<u8>", "u8")
+    .registerVectorType("vector<vector<u8>>", "vector<u8>")
+    .registerAddressType("ObjectID", 20)
+    .registerAddressType("SuiAddress", 20)
+    .registerType(
+      "ObjectDigest",
+      (writer, str) => {
+        let bytes = Array.from(decodeStr(str, 'base64'));
+        return writer.writeVec(bytes, (writer, el) => writer.write8(el));
+      },
+      (reader) => {
+        let bytes = reader.readVec((reader) => reader.read8());
+        return encodeStr(new Uint8Array(bytes), 'base64');
+      }
+    )
+);
+
+export type SuiObjectRef = {
+  objectId: string;
+  version: number;
+  digest: string;
+}
+
+initializers.push((b: typeof BCS) =>
+  b.registerStructType("SuiObjectRef", {
+    objectId: "ObjectID",
+    version: "u64",
+    digest: "ObjectDigest",
+  })
+);
+
+/**
+ * Transaction type used for transfering Coin objects between accounts.
+ * For this transaction to be executed, and `SuiObjectRef` should be queried
+ * upfront and used as a parameter.
+ */
+export type TransferCoinTx = {
+  TransferCoin: {
+    recipient: string;
+    object_ref: SuiObjectRef;
+  };
+}
+
+initializers.push((b: typeof BCS) =>
+  b.registerStructType("TransferCoinTx", {
+    recipient: "SuiAddress",
+    object_ref: "SuiObjectRef",
+  })
+);
+
+/**
+ * Transaction type used for publishing Move modules to the Sui.
+ * Should be already compiled using `sui-move`, example:
+ * ```
+ * $ sui-move build
+ * $ cat build/project_name/bytecode_modules/module.mv
+ * ```
+ * In JS:
+ * ```
+ * let file = fs.readFileSync('./move/build/project_name/bytecode_modules/module.mv');
+ * let bytes = Array.from(bytes);
+ * let modules = [ bytes ];
+ *
+ * // ... publish logic ...
+ * ```
+ *
+ * Each module should be represented as a sequence of bytes.
+ */
+export type PublishTx = {
+  Publish: {
+    modules: Iterable<Iterable<number>>;
+  };
+}
+
+initializers.push((b: typeof BCS) =>
+  b.registerStructType("PublishTx", {
+    modules: "vector<vector<u8>>",
+  })
+);
+
+// ========== Move Call Tx ===========
+
+/**
+ * An argument for the transaction. It is a 'meant' enum which expects to have
```

Comment:
what does this mean ? unsure whether users will commonly be able to interpret this